tests/run/cpp_extern.srctree

```diff
@@ -0,0 +1,148 @@
+# mode: run
+# tag: cpp
+# ticket: 1839
+
+"""
+PYTHON setup.py build_ext --inplace
+PYTHON -c "from foo import test; test()"
+PYTHON -c "from bar import test; test()"
+PYTHON -c "from baz import test; test()"
+"""
+
+######## setup.py ########
+
+from Cython.Build import cythonize
+from Cython.Distutils.extension import Extension
+from distutils.core import setup
+
+foo = Extension(
+    "foo",
+    ["foo.pyx", "foo1.cpp", "foo2.cpp"],
+)
+bar = Extension(
+    "bar",
+    ["bar.pyx", "bar1.c", "bar2.cpp"],
+)
+baz = Extension(
+    "baz",
+    ["baz.pyx", "baz1.c", "baz2.cpp"],
+    define_macros = [("__PYX_EXTERN_C", 'extern "C"')],
+)
+
+setup(
+    ext_modules=cythonize([foo, bar, baz]),
+)
+
+######## foo.pyx ########
+
+# distutils: language = c++
+
+from libcpp cimport vector
+
+cdef public vector.vector[int] get_vector():
+    return [1,2,3]
+
+cdef extern from "foo_header.h":
+    cdef size_t size_vector1()
+    cdef size_t size_vector2()
+
+def test():
+    assert size_vector1() == 3
+    assert size_vector2() == 3
+
+######## foo_header.h ########
+
+size_t size_vector1();
+size_t size_vector2();
+
+######## foo1.cpp ########
+
+#include <vector>
+#include "foo.h"
+
+size_t size_vector1() {
+    return get_vector().size();
+}
+
+######## foo2.cpp ########
+
+#include <vector>
+extern "C" {
+#include "foo.h"
+}
```

Comment:
What's the `extern "C"` here testing? I'm just wondering if it's really a C language feature and perhaps isn't something that Cython needs to test

Reply:
It tests that the `cdef public` functions are flagged as `extern "C++"` (and not `extern`) even when in a C++ compilation unit.
The rationale is that it is legal to `#include` a header from within a `extern "C"` block. In this scenario, all declarations of the header are seen with C-linkage, except those explicitly flagged `extern "C++"`.

This usage of `extern "C"` `#include` is probably a niche, yet it is legal, so I felt that Cython-generated code should comply with it, and compliance should be tested.

Reply:
Fair enough - that makes sense. It seems worth a comment since it isn't obvious at first look (at least to me)

Reply:
Comment added in commit [79f8003](https://github.com/cython/cython/pull/5040/commits/79f800360b25b8820329627d992b4b777daabdae)